driver-core/src/main/java/com/datastax/driver/core/PagingState.java

```diff
@@ -0,0 +1,158 @@
+ *
+ * PagingState comes with a certified hash verification which allows to certify the validity
+ * of the PagingState buffer. Hence, using a saved PagingState with another Statement than the
+ * one it was made with, will fail the verification and the saved PagingState won't be
+ * usable. Also, if the PagingState content buffer (or the String the PagingState was converted to)
+ * is modified , the verification will fail, too. The content of the PagingState should not be modified
+ * at any time.
+ *
+ */
+
+public class PagingState implements Serializable {
+
+    private static final long serialVersionUID = 7526471155622776147L;
+
+    private ByteBuffer pagingState;
+    private ByteBuffer hashDigest;
+
+    PagingState(ByteBuffer pagingState, Statement statement) {
+        this.pagingState = pagingState;
+        this.hashDigest = ByteBuffer.wrap(generatePagingStateHash(statement));
+    }
+
+    PagingState(ByteBuffer pagingState, ByteBuffer hashDigest) {
+        this.pagingState = pagingState;
+        this.hashDigest = hashDigest;
+    }
+
+    private byte[] generatePagingStateHash(Statement statement) {
+        byte[] digest;
+        ByteBuffer[] values;
+        MessageDigest md;
+
+        assert !(statement instanceof BatchStatement);
+
+        try {
+            md = MessageDigest.getInstance("MD5");
+            if (statement instanceof BoundStatement) {
+                BoundStatement bs = ((BoundStatement)statement);
+                md.update(bs.preparedStatement().getQueryString().getBytes());
+                values = bs.values;
+            } else {
+                //it is a RegularStatement since Batch stataments are not allowed
+                RegularStatement rs = (RegularStatement)statement;
+                md.update(rs.getQueryString().getBytes());
+                values = rs.getValues();
+            }
+            if (values != null) {
+                for (ByteBuffer value : values) {
+                    md.update(value.duplicate());
+                }
+            }
+            md.update(this.pagingState.duplicate());
+            digest = md.digest();
+
+        } catch (NoSuchAlgorithmException e) {
+            throw new RuntimeException("MD5 doesn't seem to be available on this JVM", e);
+        }
+        return digest;
+    }
+
+    boolean matches(Statement statement) {
+        ByteBuffer hashToCheck = this.hashDigest.duplicate();
+        byte[] toTest;
+        byte[] original = new byte[hashToCheck.remaining()];
+        hashToCheck.get(original);
+        toTest = this.generatePagingStateHash(statement);
+
+        return Arrays.equals(original, toTest);
+    }
+
+    /**
+     * Return the hash digest generated after the creation of the
+     * PagingState object.
+     *
+     * @return the MD5 hash generated wrapped in a {@link java.nio.ByteBuffer}.
+     */
+    public ByteBuffer getHashDigest() {
```

Comment:
Sorry for changing my mind this late, but I think exposing the two ByteBuffers separately is an implementation leak. The hash is an internal detail, regular users should not be aware of it.
So I think what we could do is use a single ByteBuffer containing the length of the paging state (1 or 2 bytes depending on maximum possible), the pagingstate, the length of the hash and the hash.
This would simplify the toString/fromString logic as well.